cli/src/cmd/cast/storage.rs

```diff
@@ -0,0 +1,246 @@
+    utils::consume_config_rpc_url,
+};
+use cast::Cast;
+use clap::Parser;
+use contract::ContractMetadata;
+use errors::EtherscanError;
+use ethers::{
+    etherscan::Client,
+    prelude::*,
+    solc::artifacts::{Optimizer, Settings},
+};
+use eyre::{ContextCompat, Result};
+use foundry_common::{
+    compile::{compile, suppress_compile},
+    try_get_http_provider,
+};
+use foundry_config::Config;
+use semver::Version;
+use std::{future::Future, pin::Pin};
+
+/// The minimum Solc version for outputting storage layouts.
+///
+/// https://github.com/ethereum/solidity/blob/develop/Changelog.md#065-2020-04-06
+const MIN_SOLC: Version = Version::new(0, 6, 5);
+
+#[derive(Debug, Clone, Parser)]
+pub struct StorageArgs {
+    // Storage
+    #[clap(help = "The contract address.", parse(try_from_str = parse_name_or_address), value_name = "ADDRESS")]
+    address: NameOrAddress,
+    #[clap(help = "The storage slot number (hex or decimal)", parse(try_from_str = parse_slot), value_name = "SLOT")]
+    slot: Option<H256>,
+    #[clap(long, env = "ETH_RPC_URL", value_name = "URL")]
+    rpc_url: Option<String>,
+    #[clap(
+        long,
+        short = 'B',
+        help = "The block height you want to query at.",
+        long_help = "The block height you want to query at. Can also be the tags earliest, latest, or pending.",
+        parse(try_from_str = parse_block_id),
+        value_name = "BLOCK"
+    )]
+    block: Option<BlockId>,
+
+    // Etherscan
+    #[clap(long, short, env = "ETHERSCAN_API_KEY", help = "etherscan API key", value_name = "KEY")]
+    etherscan_api_key: Option<String>,
+    #[clap(
+        long,
+        visible_alias = "chain-id",
+        env = "CHAIN",
+        help = "The chain ID the contract is deployed to.",
+        default_value = "mainnet",
+        value_name = "CHAIN"
+    )]
+    chain: Chain,
+
+    // Forge
+    #[clap(flatten)]
+    build: build::CoreBuildArgs,
+}
+
+impl StorageArgs {
+    pub async fn run(self) -> Result<()> {
+        let StorageArgs { address, block, build, rpc_url, slot, chain, etherscan_api_key } = self;
+
+        let rpc_url = consume_config_rpc_url(rpc_url);
+        let provider = try_get_http_provider(rpc_url)?;
+
+        let address = match address {
+            NameOrAddress::Name(name) => provider.resolve_name(&name).await?,
+            NameOrAddress::Address(address) => address,
+        };
+
+        // Slot was provided, perform a simple RPC call
+        if let Some(slot) = slot {
+            let cast = Cast::new(provider);
+            println!("{}", cast.storage(address, slot, block).await?);
+            return Ok(())
+        }
+
+        // No slot was provided
+        // Get deployed bytecode at given address
+        let address_code = provider.get_code(address, block).await?;
+        if address_code.is_empty() {
+            eyre::bail!("Provided address has no deployed code and thus no storage");
```

Comment:
Technically if the contract selfdestructed, then it could have no code but have storage. I don't think there would be any way to get the storage reliably though, geth does have a RPC endpoint `debug_storageRangeAt` but its not part of the official json rpc spec

Reply:
You're right, there's also `parity_listStorageKeys`, we could check if they are present and throw when they're not